Most active users ui

VasilyShelkov opened the pull request:
fixes #77
- so, there's 1 reducer now called mostFrequent and it is a combined reducer of both words/users
- There's a MostActiveUsers component, which gets the most active users sorted correctly in descending order.
- I've moved the actions from resultsActions to mostFrequentActions since it didn't have anything to do with the Feed and only touches the MostFrequent components
- The feed correctly filters both users/words with little performance impact, but feel free to suggest a better method if you have one.

file structure is now as follows to avoid confusion:
- results
  - mostfrequent
    - users - MostActiveUsers components
    - words - MostUsedWords components
    - ...MostFrequentComponents/Reducer/Action.js
  - ...As before
- ...As before

VasilyShelkov committed:
Initial Draft for MostActiveUsers component

VasilyShelkov committed:
Refactored a lot

VasilyShelkov committed:
npm run lint

VasilyShelkov commented:
@Winwardo ready for comments and merge

Winwardo commented:
![image](https://cloud.githubusercontent.com/assets/840549/14156183/f9bb2882-f6bc-11e5-96a4-3eaaeb1f91dd.png)

Is there anything that we can do about how little space there is here?

Winwardo commented:
![image](https://cloud.githubusercontent.com/assets/840549/14156202/0cd7effe-f6bd-11e5-9557-4512cb9d1269.png)

I like this however, it doesn't actually show the most used words for each user? Or is that planned to be added later and this is a placeholder for now?

EDIT: As per #97 this is clearly a placeholder. You may ignore this comment

Winwardo commented on src/client/results/mostfrequent/mostFrequentActions.js:
```diff
@@ -0,0 +1,23 @@
+  filterTerm,
+});
+
+export const UPDATE_MOST_ACTIVE_USERS_SEARCH_FILTER = 'UPDATE_MOST_ACTIVE_USERS_SEARCH_FILTER';
```
:+1: glad to see this as a const now, rather than how I'd made it get dynamically built before

Winwardo commented on src/client/results/mostfrequent/mostFrequentReducer.js:
```diff
@@ -0,0 +1,56 @@
+  }
+};
+
+const toggleArrayElement = (array, element) => {
```
We use this enough that it should be moved into either it's own file, or just shared/utilities.js

Winwardo replied:
And by "this" I'm referring specifically to lines 44 - 47 where we immutably slice out an item from the array

VasilyShelkov replied:
I thought about that but couldn't remember where else we used it X) but will look tomorrow

Winwardo commented on src/client/results/mostfrequent/users/MostActiveUsers.js:
```diff
@@ -0,0 +1,25 @@
+  const filteredItems = userInfoList.filter(
+    (userInfo) => userInfo.author.name.toLowerCase().includes(filterTerm.toLowerCase())
+    || userInfo.author.handle.toLowerCase().includes(filterTerm.toLowerCase())
+  ).slice(0, 100);
```
I'd consider adding explicit curly braces here as such:

```
const filteredItems = userInfoList.filter((userInfo) => {
    const matchName = userInfo.author.name.toLowerCase().includes(filterTerm.toLowerCase());
    const matchHandle = userInfo.author.handle.toLowerCase().includes(filterTerm.toLowerCase());

    return matchName || matchHandle;
})
```

to make it that bit clearer. Of course that means we ALWAYS evaluate the handle check, rather than only if the name check fails. Just a consideration, your call on which to use.

Winwardo commented on src/client/results/mostfrequent/users/UserItemsList.js:
```diff
@@ -0,0 +1,69 @@
+      </div>
+    );
+  }
+}
```
Do you think it matters that MostUsedWords is built with nested `<div>`s rather than a table? It's inconsistent between the two, but I don't think we need to worry about that at all

VasilyShelkov replied:
yer I don't think it matters really X) and swear it's using a table as well... ?

Winwardo replied:
Nope. But now we both know

VasilyShelkov replied:
you are using a table 

```
<table className="ui very basic celled table">
        <tbody>
...
```

I got that from conflated word item component in WordItemList

Winwardo commented on src/client/results/mostfrequent/users/UserItemsList.js:
```diff
@@ -0,0 +1,69 @@
+let UserItem = ({ dispatch, userInfo }) => (
+  <tr>
+    <td className="right aligned">
+      <img className="ui avatar image" src="" />
```
Please add an `alt="${userInfo.name}'s Twitter profile picture"` or similar so we're W3C compliant

VasilyShelkov commented:
What size screen you using for it to come up like that ? I mean I'm using the grid system... thought it would have downscaled but guess not :S 

I'll open up another issue to make it more responsive and fit better laters since I don't seem to have any problems with it and being responsive isn't the biggest priority right now X)

VasilyShelkov committed:
Merge with master

VasilyShelkov committed:
Fixed PR comments

- Moved ToggleArrayElement into sharedUtilities
- changed the MostActiveUser's filter criteria into variables to be more readable and explicit
- added img alt for the profile pictures in the MostActiveUser's section

VasilyShelkov commented:
@Winwardo completed your comments and all works :)
